### PR TITLE
QUIC stream limits

### LINKED ISSUE #32079: [QUIC] Decide how to handle QUIC stream limits
How should we be validating these limits?

https://github.com/dotnet/runtime/blob/eab397d509a8894da8a4baf42e0ac63a235f1e2f/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicClientConnectionOptions.cs#L29-L43

How do we get notified that a connection's stream limit has changed(MAX_STREAMS frame event)? This is required for HTTP/3

https://github.com/dotnet/runtime/blob/eab397d509a8894da8a4baf42e0ac63a235f1e2f/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs#L89-L97

https://github.com/dotnet/runtime/blob/eab397d509a8894da8a4baf42e0ac63a235f1e2f/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs#L235-L256

### PR DESCRIPTION
**Pull request by ManickaP**
Implements the 3rd option **Allowing the caller to perform their own wait** from https://github.com/dotnet/runtime/issues/32079#issuecomment-790488434

- Adds `WaitForAvailable(Bidi|Uni)rectionalStreamsAsync`:
  - triggered by peer announcement about new streams (`QUIC_CONNECTION_EVENT_TYPE.STREAMS_AVAILABLE`)
  - if the connection is closed/disposed, the method throws `QuicConnectionAbortedException` which fitted our H3 better than boolean (can be changed)
- ~Makes `Open(Bidi|Uni)directionalStreamAsync`:~
  - ~asynchronous, awaiting msquic event `QUIC_STREAM_EVENT_TYPE.START_COMPLETE`~
  - ~throwing if there's no stream capacity `QUIC_STREAM_START_FLAGS.FAIL_BLOCKED`~
  - ~immediately notifying the peer about stream being opened `QUIC_STREAM_START_FLAGS.IMMEDIATE`~
- Changes stream limit type to `int`

Fixes #32079

Note: working on adding more tests, opening PR to kick off discussion.

### CODE REVIEW

**Comment by geoffkizer** [comment]
> immediately notifying the peer about stream being opened QUIC_STREAM_START_FLAGS.IMMEDIATE

Why do this? Seems like this would be bad for H3 perf.

File: src/libraries/System.Net.Quic/ref/System.Net.Quic.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Quic/ref/System.Net.Quic.cs:30]: Why was this changed? Is it limited to int.Max by QUIC itself?

File: src/libraries/System.Net.Quic/ref/System.Net.Quic.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Quic/ref/System.Net.Quic.cs:107]: not relevant to this PR, but... weren't we going to get rid of ShutdownWriteCompleted?

File: src/libraries/System.Net.Quic/ref/System.Net.Quic.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Quic/ref/System.Net.Quic.cs:108]: What's this for?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/ResettableCompletionSource.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/ResettableCompletionSource.cs:13]: You unsealed this, but I don't actually see where you are deriving from it? Did I miss something?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:54]: typo: `ressetable` => `resetable`

Comment (geoffkizer) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:54]: Also nit: we don't need to reallocate "every time we set the result", right? We just can't reuse them.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:56]: "keep the TCSes null" is a bit awkward, I would suggest "allocate when needed"

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:197]: Probably want to set these to null just in case.

**Comment by geoffkizer** [comment]
> throwing if there's no stream capacity QUIC_STREAM_START_FLAGS.FAIL_BLOCKED

If this is the case, then why does `Open*Stream` need to be async? Wouldn't it just succeed or fail immediately?

**Comment by ManickaP** [comment]
> > immediately notifying the peer about stream being opened QUIC_STREAM_START_FLAGS.IMMEDIATE
> 
> Why do this? Seems like this would be bad for H3 perf.

This might be my misunderstanding from our discussions. I recall (possibly wrongly) that @scalablecory was surprised we weren't using it. The only drawback I see is that without this flag, the peer might not know about the number of streams reserved (if they do not start sending data immediately) and as a result not sending new MAX_STREAMS ahead of time. However, I don't have a strong preference here.

---
>> throwing if there's no stream capacity QUIC_STREAM_START_FLAGS.FAIL_BLOCKED
>
> If this is the case, then why does `Open*Stream` need to be async? Wouldn't it just succeed or fail immediately?

Msquic will always send `QUIC_STREAM_EVENT_START_COMPLETE` asynchronously, we obviously don't need to wait for it in `Open*Stream` methods. That will postpone the waiting till the first data send or at least that's the way it's programmed ATM. If I were a consumer of Quic APIs, I'd expect such behavior, but I do concur that it's not optimal performance-wise.
Lastly why the method is async, that's something I think we should stick with even if we decide to implement it synchronously. This will give us much more leeway if we decide to change (e.g. to not use `QUIC_STREAM_START_FLAG_FAIL_BLOCKED`) or even swap the implementation (e.g. the fully managed QUIC) and the overhead of `ValueTask` doesn't seem such a high price. But since we're not shipping APIs in 6.0, we might postpone this decision and keep it sync.

This is great discussion, we should revisit some of the topics in our sync meeting, cc: @scalablecory @CarnaViire @wfurt

File: src/libraries/System.Net.Quic/ref/System.Net.Quic.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/ref/System.Net.Quic.cs:30]: Msquic returns `ushort`:
https://github.com/dotnet/runtime/blob/51b6188ed36fea7ad03e24a36224a600cdad1c72/src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs#L338-L346
AFAIK, we usually go with signed types for APIs ==> `int`.

I'm aware that QUIC sends `long`-range values in [MAX_STREAMS frame](https://quicwg.org/base-drafts/rfc9000.html#name-max_streams-frames), but that's cumulative number of streams which msquic hides from us. This is probably the original motivation for the `long` type.

Once again, I'm not against keeping it `long`. We discussed it briefly in one of our meetings and the consensus was for `int` so went ahead with the change.

File: src/libraries/System.Net.Quic/ref/System.Net.Quic.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/ref/System.Net.Quic.cs:107]: I think it should have been resolved with the `DuplexStream` API, but that didn't pan out AFAIK. We should reopen the discussion about this.

cc: @scalablecory

File: src/libraries/System.Net.Quic/ref/System.Net.Quic.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/ref/System.Net.Quic.cs:108]: For `QUIC_STREAM_EVENT_START_COMPLETE` notification. This will either survive or not depending how we decide to handle the stream opening, as described in https://github.com/dotnet/runtime/pull/52704#issuecomment-847089688

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/ResettableCompletionSource.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/ResettableCompletionSource.cs:13]: This is a mistake, thanks for noticing.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:199]: My original thinking was that the exception is the final state of the TCS and than is somebody calls `WaitForAvailable*StreamsAsync`, they would get this exception.
However, I already have a check in place for that:
https://github.com/ManickaP/runtime/blob/78b526da7f2460f90b7936e7044f5519ef8312f8/src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs#L406-L409

So I decided to go with the suggestion.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs]: Fixed, reworded.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs]: Fixed, reworded.

**Comment by ManickaP** [comment]
So we do the hack for 0-byte write before getting the streamId:
https://github.com/dotnet/runtime/blob/57bdb959be336482dfb6060ef62f935c6fecee3f/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs#L215-L217

But it seems like it's not necessary. I was able to get the stream id right after calling `QuicStreamStart` which also confirms what can be found in msquic logs and code:
https://github.com/microsoft/msquic/blob/a38f8c5399917c3921be6e52b870cff4099e10a5/src/core/stream.c#L574-L596

So I'll get rid of the hack as well as the QUIC_STREAM_START_FLAGS.IMMEDIATE.

File: src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs
Comment (ManickaP) [inline src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs:57]: These 2 new tests pass on their own but if run within the whole suite they cause the crash with stream being finalized after the connection. That should be fixed with https://github.com/dotnet/runtime/pull/52800, I'll revisit it enabling them afterwards.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs
Comment (CarnaViire) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs:168]: Should we possibly fail fast from the start, if `_connection` is null? We wouldn't wait to acquire lock around `_activeRequests` down there (line 185) that we don't actually need in this case, and it will also reduce block nesting depth (or whatever it's called).

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs
Comment (CarnaViire) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:250]: would it be beneficial to do both uni and bidi updates inside a single lock, not in two subsequent?
```suggestion
            TaskCompletionSource? unidirectionalTcs;
            TaskCompletionSource? bidirectionalTcs;
            lock (state)
            {
                if (connectionEvent.Data.StreamsAvailable.UniDirectionalCount > 0)
                {
                    unidirectionalTcs = state.NewUnidirectionalStreamsAvailable;
                    state.NewUnidirectionalStreamsAvailable = null;
                }
                if (connectionEvent.Data.StreamsAvailable.BiDirectionalCount > 0)
                {
                    bidirectionalTcs = state.NewBidirectionalStreamsAvailable;
                    state.NewBidirectionalStreamsAvailable = null;
                }
            }
            if (unidirectionalTcs is not null)
            {
                unidirectionalTcs.SetResult();
            }
            if (bidirectionalTcs is not null)
            {
                bidirectionalTcs.SetResult();
            }
```

File: src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs
Comment (CarnaViire) [reply src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs:57]: We discussed offline that it is caused by not accepted stream getting stuck inside accept queue and not being disposed there. You can try accepting it (after all test checks if it's important for the test) and disposing manually, or you may wait for Tomas' fix, your choice.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs
Comment (ManickaP) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs:168]: What I have here is actually wrong, we should check the `_connection` inside the lock. We handle all the changes to it inside a lock. So I have a potential race here. I need to fix this.
Thanks for pointing at this code and the locking!

Comment (ManickaP) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs]: I also reorder the statements here, so your point is addressed.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:250]: Anyone else has an opinion on this? Will single lock be more efficient here?

File: src/libraries/System.Net.Http/src/System.Net.Http.csproj
Comment (CarnaViire) [inline src/libraries/System.Net.Http/src/System.Net.Http.csproj:622]: Should that change be reverted?

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs
Comment (CarnaViire) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs:171]: Here and other `Console.WriteLine`s - I guess it should be reverted as well

File: src/libraries/System.Net.Http/src/System.Net.Http.csproj
Comment (ManickaP) [reply src/libraries/System.Net.Http/src/System.Net.Http.csproj:622]: Yes all the changes from the last commit are to be reverted. This one included.
All the Console.WriteLines will go away. you can ignore them.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs
Comment (ManickaP) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs:171]: Same as above.

**Comment by CarnaViire** [comment]
LGTM, waiting on test cleanup and resolving a question about two locks vs one

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs]: I decided to go with the suggestion and not wait with it.

**Comment by ManickaP** [comment]
I moved the experiments with the failing test into its own branch and disabled it in this PR (with an ActiveIssue).
Everything should be cleaned up a addressed now. Only waiting for a final review.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs
Comment (CarnaViire) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:205]: I remembered that in Quic we always had SetException and SetResult outside of the lock:
https://github.com/dotnet/runtime/blob/5fc93f425ead47546efcbe8fa1be0b1becc03c25/src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs#L394-L407
I wonder why is it like that and whether that's a best practice? @scalablecory @geoffkizer

**Review by CarnaViire — APPROVED** [review]
LGTM, but it would be great if @scalablecory or @geoffkizer could also take a look.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs]: I pulled the `SetException` outside of the lock. What is important is that the swaps of the variables stay under the lock, the set itself doesn't have to be.

**Pull request merged**
